plutus-core/executables/Common.hs

```diff
@@ -355,7 +355,7 @@ writeFlat ::
 writeFlat outp flatMode prog = do
   flatProg <- case flatMode of
             Named         -> pure $ serialiseProgramFlat (() <$ prog) -- Change annotations to (): see Note [Annotation types].
-            DeBruijn      -> typedDeBruijnNotSupportedError -- TODO, this should work but we don't have a program here :/
+            DeBruijn      -> serialiseDbProgramFlat (() <$ prog)
```

Comment:
Wait!  This works for untyped programs, but I think this code also handles typed ones and I'm not sure what'll happen then.

Reply:
That seems to be OK.  The `uplc` command can convert the program below from PLC source to flat, but `pcl` fails with the "De-Bruijn-named ASTs are not supported for typed Plutus Core" message.

```
(program 1.1.1
[ (builtin multiplyInteger) (con integer 111) (con integer 222) ]
)
```

Reply:
No, now I'm confused again, Aren't the `DeBruijn` and `NamedDeBruijn` cases doing exactly the same thing?

Reply:
There's some rather confusing ad-hoc polymorphism going on here, I'm quite confused about what is going on also.

Reply:
I've worked out what's going on now and I think that `writeFlat` is doing the wrong thing.  We have at least three kinds of untyped ASTs, which use different types of Names:

  * `Named`: these have names containing both Uniques and textual identifiers.
  * `DeBruijn`: these have names which are just De Bruijn indices (no text), and are what we usually use for serialsation.
  * `NamedDeBruijn`: names are De Bruijn indices along with textual identifiers.
  
I'm not sure what `NamedDeBruijn` is for, despite the fact that it was probably me that added it (maybe for consistency with the CBOR stuff that we used to have).

Anyway, reading a program is handled by [`loadUplcASTfromFlat`](https://github.com/input-output-hk/plutus/blob/c1519ca8f65c8a87795d92b8c5a36a30047cbbd6/plutus-core/executables/Common.hs#L316).  It's given a parameter (ultimately coming from the `--if` option) which tells it what kind of thing to deserialise, and I think it does the right thing.  It reads in a flat file and converts it to Named format: for `Named` it just deserialises, for `NamedDeBruijn` it deserialises and applies `fromDeBruijn`, retaining the textual IDs, and for `DeBruijn` it deserialises, adds an "x" identifier to each name, and then applied `fromDeBruijn`.

However, writing a program is handled by `writeFlat` above.  For `Named` serialisation it calls the `serialiseProgramFlat` _function_, which does the right thing, but for `DeBruijn` (after my edit) and `NamedDeBruijn` it calls the `serialiseDbProgramFlat` _typeclass method_ for the `UplcProgram` type, which is implemented [here](https://github.com/input-output-hk/plutus/blob/c1519ca8f65c8a87795d92b8c5a36a30047cbbd6/plutus-core/executables/Common.hs#L96). That calls `toDeBruijn` (which throws away the textual IDs and then does the De Bruijn transformation) and then serialises, so it's always saving things in un-named De Bruijn indexed form, whereas for `NamedDeBruijn` we want to preserve the textual IDs.

This is all very confusing! I think there are at least three things we can do: 
  * Get rid of the `serialiseDbProgramFlat` method and use explicit functions instead.
  * Add an `AstNameType` parameter to `serialiseNamedDbProgramFlat`, which would make it consistent with the `loadASTfromFlat` method.
  * Do away with `NamedDeBruijn` serialisation altogether (and maybe even `Named` serialisation).
  
I'm not sure which of these is best. I'll think about it and try to remember why we have `Named` and `NamedDeBruijn` serialisation in the first place. Maybe these are useful for programs which analyse PLC scripts, since they contain more information than the simple De Bruijn indexed form.

Reply:
Anyway, I think this PR solves my original problem.  However, `--of flatNameDeBruijn` does the wrong thing, as it did before.

Reply:
- when I refactored this I tried to keep the meaning of the code the same. I'm not sure if I succeeded. It may be the case that I didn't succeed.
- I do remember the not supported error though for `DeBruijn`. @kwxm do you remember maybe there was a reason we didn't want to support that?
- I like to get rid of things that we don't need. So I vote to do away with `NamedDeBruijn` if we're not using it. It can be in a follow up PR but let's fix things when we notice the problem, so we don't get confused again.

Reply:
> I do remember the not supported error though for DeBruijn. @kwxm do you remember maybe there was a reason we didn't want to support that?

No, I don't remember.  I do want to support it though, since that's the format we use for transmission to the chain and it's useful to convert to and from that format.  I think I can cope without `NamedDeBruijn` though, although I'll think a bit more about that first. There's still a `flat` instance for that, so we can put it back if necessary.

I think it would make things a bit more consistent if we added  an `AstNameType` parameter to `serialiseNamedDbProgramFlat`, like `loadASTfromFlat`.  

If you're OK with those changes I could add them to this PR: I don't think it'd be too difficult.

Reply:
Opps, just seen @michaelpj's reply.  In that case I vote to keep `NamedDeBruijn`, but we should add the `AstNameType` parameter to the serialisation method to make it easy to output that.

Reply:
>  In that case I vote to keep NamedDeBruijn, but we should add the AstNameType parameter to the serialisation method to make it easy to output that.

I've been wondering whether we should do something like what `cardano-api` does, where it has a JSON "envelope" format, which is just a two-field object where the first field tells you what "type" of thing it is, while the second field contains the serailized data. Then we could have typed input formats, allowing the commands to largely just DTRT, without having to put more tags into the "real" serialization format...

Reply:
> a JSON "envelope" format

But wouldn't the flat-encoded stuff have to be encoded in some textual form then, making it significantly bigger?